src/java.desktop/share/classes/javax/swing/text/AbstractDocument.java

```diff
@@ -2320,9 +2320,9 @@ public Element positionToElement(int pos) {
          */
         public void replace(int offset, int length, Element[] elems) {
             int delta = elems.length - length;
-            int src = offset + length;
-            int nmove = nchildren - src;
-            int dest = src + delta;
+            int src = Math.abs(offset + length);
+            int nmove = Math.abs(nchildren - src);
+            int dest = Math.abs(src + delta);
```

Comment:
I'm rather unsure why any of these could become negative. Would it be possible that using `Math.abs` corrupts the document or leads to a data loss?

The thing is the `UndoableEdit` was produced while the document and the component didn't have `bidi` set to `true`. Inserting a right-to-left text changes that situation, and `bidiRoot` starts to contain elements. I assume changing the component orientation also modifies the contents of `bidiRoot`. As such, performing the undo operation may become impossible because the recorded state in the `UndoableEdit` is inconsistent with the current state of the document.